sema/subyt/api.py

```diff
@@ -0,0 +1,379 @@
+import itertools
+import logging
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from typing import Callable, Dict
+
+from sema.commons.log.loader import load_log_config
+
+load_log_config()
+log = logging.getLogger(__name__)
+
+
+class Sink(ABC):
+    """Abstract Interface for sinks"""
+
+    def __init__(self) -> None:
+        # lastModifiedTime for each file in the sink, t = 0 by default
+        self.mtimes = {".": 0.0}
+
+    @abstractmethod
+    def open(self):
+        """Open file handle to Sink"""
+
+    @abstractmethod
+    def close(self):
+        """Notifies the Sink all writing has been done - allows cleanup"""
+
+    @abstractmethod
+    def add(self, part: str, item: dict = None, source_mtime: float = None):
+        """Writes out the part to the sink
+
+        :param part: the result for a specific part that needs to be sinked
+        :type part: str
+        :param item: the record for which this part was produced
+        :type item: dict
+        """
+
+
+class Source(ABC):
+    """Abstract Interface for input Sources - mainly context-manager
+    for the actual Iterable to be returned"""
+
+    def __init__(self) -> None:
+        # lastModifiedTime for each file in the source, t = infinity by default
+        self.mtimes = {".": float("inf")}
+
+    @abstractmethod
+    def __enter__(self) -> Iterable:
+        """Source context is expected to yield an Iterable"""
+
+    @abstractmethod
+    def __exit__(self, *exc):
+        """Source context cleanup"""
+
+
+# TODO make a pandas source in sources.py
```

Comment:
remove and replace by having a todo-issue for this

Reply:
taken up in #70